### ISSUE TITLE
Rule Change: support more binary operators

### ISSUE THREAD

**Issue by camc314** [eslint/eslint#20573]
### What rule do you want to change?

no-constant-binary-expressions

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
const shouldShowStreak = profile.streak ?? 0 > 1
```

Currently doesn't report an error, despite `0 > 1` always evaluating to false. Since this is a constant binary expression, I think makes semi

### What does the rule currently do for this code?

Reports no error:

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tY29uc3RhbnQtY29uZGl0aW9uOiBcImVycm9yXCIsIFxuICAgICAgICAgbm8tY29uc3RhbnQtYmluYXJ5LWV4cHJlc3Npb246IFwiZXJyb3JcIiAqL1xuXG5jb25zdCBzaG91bGRTaG93U3RyZWFrID0gcHJvZmlsZS5zdHJlYWsgPz8gMCA+IDE7XG5cbmNvbnN0IGFsd2F5c1RydWUgPSAxID4gMDtcblxuXG5pZiAodHJ1ZSkge31cbmNvbnN0IHggPSAxID09PSAyOyIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

### What will the rule do after it's changed?

Report an error on code such as 

```
const shouldShowStreak = profile.streak ?? 0 > 1
```

Explaining that `0 > 1` is a binary expression that always evaluates to the same value.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

Related:

https://github.com/oxc-project/oxc/issues/19825
https://github.com/oxc-project/oxc/issues/19847

**Comment by loontrix** [comment]
More cases from https://github.com/oxc-project/oxc/issues/19847:

- Fails:
  - `const a = 1 > 0`
  - `const b = 0 > 1`
  - `const shouldShowStreak = profile.streak ?? 0 > 1`
- Passes:
  - `const a = x > 0`
  - `const a = foo() > 0`
  - expressions with possible side effects in either operand

I can work on this.

**Comment by DMartens** [comment]
I agree that the rule should support the binary operators `<`, `<=`, `>` and `>=`.
As this is a rule change, we require multiple team members to support this proposal.

**Comment by bradzacher** [comment]
Another case not listed is the equality check operator.

For example it doesn't catch `1 === 2`

**Comment by JoshuaKGoldberg** [comment]
I'm in favor. This seems like a clear gap.

**Comment by lumirlumir** [comment]
Friendly ping @eslint/eslint-team for consensus on this suggestion.

**Comment by fasttime** [comment]
If `no-constant-binary-expressions` is extended to report additional problems, I think it should be behind a new option to avoid a breaking change for existing user.

One other consideration is overlap with `no-constant-condition`, which already reports constant comparisons when they appear in control-flow conditions, e.g.:

```js
/* eslint no-constant-condition: "error" */

if (0 > 1) {
    // ...
}
```

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWNvbnN0YW50LWNvbmRpdGlvbjogXCJlcnJvclwiICovXG5cbmlmICgwID4gMSkge1xuICAgIC8vIC4uLlxufVxuIiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

That said, there's already some overlap today in cases like `if (false && foo)`, so this may be something to address separately.

Apart from that, I'd be in favor of extending the rule to report on comparisons when both operands are literals. That seems useful.

**Comment by WajihaNiazi** [comment]
Hi 👋
I’m interested in working on this issue. Is it currently available to be assigned?

This would be my first contribution to this repository, so I’d like to take some time to explore the codebase, run it locally, and understand the context before starting.

If it’s okay, I can take this on and aim to submit a PR within about two weeks.

Please let me know if that works or if there are any guidelines I should follow before starting.

Thanks!

**Comment by lumirlumir** [comment]
@WajihaNiazi Thanks for your interest in contributing!

However, this issue hasn't reached a consensus yet. If a consensus is reached, you'd be welcome to take it on.

**Comment by WajihaNiazi** [comment]
> [@WajihaNiazi](https://github.com/WajihaNiazi) Thanks for your interest in contributing!
> 
> However, this issue hasn't reached a consensus yet. If a consensus is reached, you'd be welcome to take it on.

Thanks for the update, really appreciate it!

That makes sense. I’ll keep an eye on the discussion and wait until a consensus is reached. I’d still be very interested in taking this on once it’s ready.

Looking forward to contributing 🚀

**Comment by DMartens** [comment]
I think there is team consensus (3 members in favor) with the constraint that both operands are literals, so marking as accepted.
@WajihaNiazi feel free to raise a PR.

**Comment by loontrix** [comment]
@DMartens I had already requested to work on this first.

**Comment by DMartens** [comment]
Oh, I am sorry @loontrix. I seem to have skipped that while skimming.

**Comment by kecrily** [comment]
Are you still working on this? friendly ping @loontrix

**Comment by sethamus** [comment]
I've been looking into this and have a fix working locally. @loontrix, if you'd prefer, I can open a PR with the changes.

**Comment by kecrily** [comment]
Let’s wait another week to see if he comes back. Otherwise, a long wait will prevent others from taking over.

**Comment by loontrix** [comment]
Sorry, I've been dealing with some health issues and haven't been able to keep up. @sethamus, please go ahead and open the PR.
